Bugfix/52688 gcp compute missing image

maxamillion opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
add boot image mapping to gcp_compute instance data for all disk
image data in the configured zones
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
gcp_compute

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

This is a continuation of the work started here: https://github.com/ansible/ansible/pull/52688

rambleraptor committed:
adding image field to inventory script

rambleraptor committed:
pylint fixes

maxamillion committed:
add boot image mapping to gcp_compute instance data for all disk
image data in the configured zones

Signed-off-by: Adam Miller <admiller@redhat.com>

AlanCoding commented:
just at a quick glance, this is returning the expected "image" field. 👍 

Performance was a concern, so let me record some timings:

branch:
2m37.724s
2m35.697s

devel:
1m18.300s

This isn't actually unexpected. I'm probably rate limited, myself. The plugin tended to use a single request, and this looks about probably 2x slower, probably due to 1 additional request for the image. The script might actually be even slower.

I haven't tested after the rebase, and I should probably make sure to double-check requests being made.

rambleraptor commented:
This looks awesome! Is there a way to put a flag in to turn off the image field (in case performance is an issue for people)

maxamillion commented:
rebased and added a changelog

maxamillion commented:
@rambleraptor I'm open to that ... should we opt-in to the `image` field or opt-out?

@bcoca @AlanCoding thoughts? ^

rambleraptor commented:
In my mind, `image` is a much more optional field. I'm don't see most users getting value out of it (please correct me if I'm wrong). I'd say opt-in, but this is a very weak opinion.

maxamillion commented:
@rambleraptor I agree wit you, it probably is ... I was just curious about opinions on this being posted to the PR for my own curiosity and posterity. I don't mind adding an opt-in toggle.

AlanCoding commented:
I rebased, hitting a new error

```
 [WARNING]:  * Failed to parse /Users/alancoding/Documents/repos/ansible-inventory-file-examples/private/gce/gcp_compute.yml with auto plugin: 'scopes'

  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 272, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/auto.py", line 58, in parse
    plugin.parse(inventory, loader, path, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/gcp_compute.py", line 429, in parse
    project_disks = self._get_project_disks(config_data, query)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/gcp_compute.py", line 361, in _get_project_disks
    params=request_params
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/module_utils/gcp_utils.py", line 86, in get
    return self.session().get(url, **kwargs)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/module_utils/gcp_utils.py", line 133, in session
    self._credentials())
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/module_utils/gcp_utils.py", line 160, in _credentials
    return service_account.Credentials.from_service_account_file(path).with_scopes(self.module.params['scopes'])
```

This is a `KeyError`, which the inventory parsing logic fails to report IIRC, I've seen that before.

rambleraptor commented:
You have to set the scopes parameter like [here](https://github.com/ansible/ansible/blob/5682ea90054238b765665de42f45708a731c0074/lib/ansible/plugins/inventory/gcp_compute.py#L420).

AlanCoding commented:
requests look like:

```
request: https://accounts.google.com/o/oauth2/token, time: 0.110888957977, size: 202
request: https://www.googleapis.com/compute/v1/projects/<project name>/aggregated/disks?filter=&maxResults=500, time: 75.7364270687, size: 84179
request: https://accounts.google.com/o/oauth2/token, time: 76.002243042, size: 202
request: https://www.googleapis.com/compute/v1/projects/<project name>/zones/us-east1-d/instances?filter=, time: 76.1251411438, size: 159766
```

Ignore the timings.

This looks relatively similar to the script, except for the extra auth request (fine by me). See https://github.com/ansible/ansible/issues/51884 for that background.

It doesn't scale with the number of hosts, and shouldn't be any slower than the script was. So that's all I was looking for.

I don't have a position on whether this should be opt-in. It's trivial for me to use the option, if you do go that route. So it doesn't matter to me.

AlanCoding commented:
In the config file I did add

```yaml
scopes:
  - https://www.googleapis.com/auth/compute
```

I don't think that's right. I didn't have to have this before, and it's redundant information in 3 different places now.

However, I don't think that was introduced with this PR.

rambleraptor commented:
Yeah, the `scopes` parameter is old tech debt. The modules add it in automatically, but that change hasn't made it over to the utility file.

maxamillion committed:
make retrieve_image_info optional, remove second auth, fix default scopes

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion commented:
@AlanCoding you no longer need to set `scopes`, that was a goof on my part when I rebased on `devel` because I didn't properly switch to the new var where the default is stored. I also removed the second auth (no idea why I did that in the first place).

bcoca commented:
i re-added scopes  as they are part of google's underlying auth and there might be cases that were using it (i.e compute.readonly)
